skimmer/skimmer_utils.py

```diff
@@ -146,45 +148,36 @@ def is_data(events):
-        phi_sub = [2.03,3.01,-1.75,-0.77,2.73,0.21,0.07,0.21,-2.59,0.77,0.91,1.75,1.75,2.87,0.63,-1.19,-2.31,-2.17,2.73,-0.77,-0.77,-3.15,-0.49]
-    elif year == "2018":
-        eta_lead = [1.488,1.488,1.584]
-        phi_lead = [-0.77,-0.63,-0.77]
-        eta_sub = [-1.584,-1.2,-0.912,-0.912,-0.816,-0.816,-0.72,-0.72,-0.528,-0.432,-0.336,-0.24,-0.24,-0.144,-0.144,-0.144,-0.048,-0.048,0.144,0.912,0.912,1.008,1.296,-0.72,1.104,1.488,1.776]
-        phi_sub = [0.63,-1.19,2.03,3.01,-1.75,-0.77,-2.17,-0.77,2.73,2.73,0.21,0.07,0.21,-2.59,0.07,0.77,0.77,0.91,1.75,1.75,2.87,0.63,-0.49,-2.31,-3.15,-0.21,0.77]
-    else:
-        raise ValueError("Invalid year")
 
-    eta_lead = nb.typed.List(eta_lead)
-    eta_sub = nb.typed.List(eta_sub)
-    phi_lead = nb.typed.List(phi_lead)
-    phi_sub = nb.typed.List(phi_sub)
+def get_phi_spike_filter(hot_spots_dict,var_name,j_eta_i,j_phi_i,rad):
+    hot_etas_i, hot_phis_i = hot_spots_dict[var_name]
+    hot_etas_i_reshaped = np.reshape(hot_etas_i,(len(hot_etas_i),1))
+    hot_phis_i_reshaped = np.reshape(hot_phis_i,(len(hot_phis_i),1))
+    j_eta_i_reshaped = np.broadcast_to(list(j_eta_i),(len(hot_etas_i),len(j_eta_i)))
+    j_phi_i_reshaped = np.broadcast_to(list(j_phi_i),(len(hot_phis_i),len(j_phi_i)))
+    return np.prod((j_eta_i_reshaped - hot_etas_i_reshaped)**2 + (j_phi_i_reshaped - hot_phis_i_reshaped)**2 > rad, axis=0, dtype=bool)
 
+
+def apply_phi_spike_filter(events, year, hot_spots_pkl, channel="t", jet_eta_branch_name="Jet_eta", jet_phi_branch_name="Jet_phi"):
+    with open(hot_spots_pkl,"rb") as infile:
+        phi_spike_hot_spots = pickle.load(infile)
+    rad = 0.028816*0.35 # the factor of 0.35 was optimized from the signal vs. background sensitivity study for s-channel
+    hot_spots_dict = phi_spike_hot_spots[year]
     jets_eta = getattr(events, jet_eta_branch_name)
     jets_phi = getattr(events, jet_phi_branch_name)
-
-    builder = ak.ArrayBuilder()
-    phi_spike_filter = __get_phi_spike_filter(builder, eta_lead, phi_lead, eta_sub, phi_sub, rad, jets_eta, jets_phi, reverse=reverse).snapshot()
-
-    events = events[phi_spike_filter]
-
+    if channel == "t":
+        pass1 = get_phi_spike_filter(hot_spots_dict,"j1Phivsj1Eta",jetVar_i(jets_eta,0),jetVar_i(jets_phi,0),rad)
+        pass2 = get_phi_spike_filter(hot_spots_dict,"j2Phivsj2Eta",jetVar_i(jets_eta,1),jetVar_i(jets_phi,1),rad)
+        pass3 = get_phi_spike_filter(hot_spots_dict,"j3Phivsj3Eta",jetVar_i(jets_eta,2),jetVar_i(jets_phi,2),rad)
+        pass4 = get_phi_spike_filter(hot_spots_dict,"j4Phivsj4Eta",jetVar_i(jets_eta,3),jetVar_i(jets_phi,3),rad)
+    elif channel == "s":
+        pass1 = get_phi_spike_filter(hot_spots_dict,"j1Phivsj1Eta",jetVar_i(jets_eta,0),jetVar_i(jets_phi,0),rad)
+        pass2 = get_phi_spike_filter(hot_spots_dict,"j2Phivsj2Eta",jetVar_i(jets_eta,1),jetVar_i(jets_phi,1),rad)
+    events = events[pass1 & pass2 & pass3 & pass4]
```

Comment:
Instead of having this, wouldn't it be more flexible to have a `n_jets` variable to define the number of jets on which the filter is applied and loop over the jets? So that everyone can freely choose depending on his/her analysis?

Reply:
Also the last line with break for the s-channel currently ;)

Reply:
Also, if `jetVar` and `get_phi_spike_filter` are internal functions only used here, can you make them private starting the function name with two underscores `__`?